File: Cracking the Code Interview/Arrays and Strings/URLify/cpp/urlify.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    public:
        void urlify(char *str, int len) {
            int numOfSpaces = 0;
            int i = 0, j = 0;
            for ( i = 0; i < len; ++i ) {
                if (str[i] == ' ') {
                    ++numOfSpaces;
                }
            }
            
            int extendedLen  = len + 2 * numOfSpaces;
            i = extendedLen - 1;
            for( j = len - 1; j >= 0; --j ) {
                if ( str[j] != ' ' ) {
                str[i--] = str[j];
                } else {
                    str[i--] = '0';
                    str[i--] = '2';
                    str[i--] = '%';
                }
            }
        }

};


int main() {

    Solution solution;
    char str[] = "Mr John Smith    ";                       //String with extended length ( true length + 2* spaces)
    cout << "Actual string   : " << str << std::endl;
    solution.urlify(str, 13);                                        //Length of "Mr John Smith" = 13
    cout << "URLified string : " << str << std::endl;
    return 0;
}

File: Cracking the Code Interview/Arrays and Strings/URLify/python/urlify.py
import unittest

def urlify_algo(s, length):
    # convert to list because Python strings are immutable
    charList = list(s)
    newIndex = len(charList)

    for i in reversed(range(length)):
        if charList[i] == " ":
            charList[newIndex-3: newIndex] = "%20"
            newIndex -= 3
        else:
            charList[newIndex-1] = charList[i]
            newIndex -= 1

    return "".join(charList[newIndex:])
    
def urlify_pythonic(s, length):
    """solution using standard library"""
    return s[:length].replace(" ", "%20")

class Test(unittest.TestCase):
    """Test Cases"""

    test_cases = {
        ("much ado about nothing      ", 22): "much%20ado%20about%20nothing",
        ("Mr John Smith       ", 13): "Mr%20John%20Smith",
        (" a b    ", 4): "%20a%20b",
        (" a b       ", 5): "%20a%20b%20",
    }
    testable_functions = [
                            urlify_algo, 
                            # urlify_pythonic
                        ]

    def test_urlify(self):
        for urlify in self.testable_functions:
            for args, expected in self.test_cases.items():
                actual = urlify(*args)
                assert actual == expected, f"Failed {urlify.__name__} for: {[*args]}"


if __name__ == "__main__":
    unittest.main()